File: Python/NumMatrix.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: NumArray.py
@Function: 区域和检索 - 数组不可变
@Link: https://leetcode-cn.com/problems/range-sum-query-immutable/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-26
"""


from typing import List


class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        m, n = len(matrix), (len(matrix[0]) if matrix else 0)
        self.sums = [[0] * (n + 1) for _ in range(m)]
        _sums = self.sums

        for i in range(m):
            for j in range(n):
                _sums[i][j + 1] = _sums[i][j] + matrix[i][j]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        _sums = self.sums

        total = sum(_sums[i][col2 + 1] - _sums[i][col1]
                    for i in range(row1, row2 + 1))
        return total


# -------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/NumMatrix.cpp
/**
 * @File    : NumMatrix.cpp
 * @Brief   : 二维区域和检索 - 矩阵不可变
 * @Link    : https://leetcode-cn.com/problems/range-sum-query-2d-immutable/
 * @Author  : Wei Li
 * @Date    : 2021-07-26
*/

#include <iostream>
#include <vector>

class NumMatrix
{
    std::vector<std::vector<int>> integral;

public:
    NumMatrix(std::vector<std::vector<int>> matrix)
    {
        int m = matrix.size(), n = m > 0 ? matrix[0].size() : 0;
        integral = std::vector<std::vector<int>>(m + 1, std::vector<int>(n + 1, 0));
        for (int i = 1; i <= m; ++i)
        {
            for (int j = 1; j <= n; ++j)
            {
                integral[i][j] = matrix[i - 1][j - 1] + integral[i - 1][j] +
                                 integral[i][j - 1] - integral[i - 1][j - 1];
            }
        }
    }
    int sumRegion(int row1, int col1, int row2, int col2)
    {
        return integral[row2 + 1][col2 + 1] - integral[row2 + 1][col1] -
               integral[row1][col2 + 1] + integral[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
